<?xml version="1.0" ?>
<robot name="terrabot_design4" xmlns:xacro="http://www.ros.org/wiki/xacro">

<xacro:include filename="$(find terrabot_design4_description)/urdf/materials.xacro" />
<xacro:include filename="$(find terrabot_design4_description)/urdf/terrabot_design4.trans" />
<xacro:include filename="$(find terrabot_design4_description)/urdf/terrabot_design4.gazebo" />

<link name="base_link">
  <inertial>
    <origin xyz="-0.002000000000000839 -0.00013843475574249804 0.19998236442836137" rpy="0 0 0"/>
    <mass value="0.048705056378264104"/>
    <inertia ixx="9.7e-05" iyy="5.9e-05" izz="0.000153" ixy="-0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/base_link.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/base_link.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="wheel3_1">
  <inertial>
    <origin xyz="-0.007499999999999951 8.259963496515255e-08 1.392678549635984e-07" rpy="0 0 0"/>
    <mass value="0.10956304379394403"/>
    <inertia ixx="0.000137" iyy="7.1e-05" izz="7.1e-05" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="-0.056 0.099795 -0.049" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/wheel3_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="-0.056 0.099795 -0.049" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/wheel3_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="wheel2_1">
  <inertial>
    <origin xyz="0.007499999999999979 8.259963550638627e-08 1.3926785466522595e-07" rpy="0 0 0"/>
    <mass value="0.10956304379394405"/>
    <inertia ixx="0.000137" iyy="7.1e-05" izz="7.1e-05" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0.067 0.099795 -0.049" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/wheel2_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="0.067 0.099795 -0.049" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/wheel2_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="wheel4_1">
  <inertial>
    <origin xyz="-0.007499999999998987 1.9826153256630175e-07 3.4161735154580875e-07" rpy="0 0 0"/>
    <mass value="0.10956304379394403"/>
    <inertia ixx="0.000137" iyy="7.1e-05" izz="7.1e-05" ixy="0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="-0.059 -0.100321 -0.049" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/wheel4_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="-0.059 -0.100321 -0.049" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/wheel4_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="wheel1_1">
  <inertial>
    <origin xyz="0.007500000000000777 1.9826153252466838e-07 3.416173510253917e-07" rpy="0 0 0"/>
    <mass value="0.10956304379394403"/>
    <inertia ixx="0.000137" iyy="7.1e-05" izz="7.1e-05" ixy="-0.0" iyz="0.0" ixz="0.0"/>
  </inertial>
  <visual>
    <origin xyz="0.064 -0.100321 -0.049" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/wheel1_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="0.064 -0.100321 -0.049" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/wheel1_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="gear2_1">
  <inertial>
    <origin xyz="0.03693017933671068 7.887551849927732e-05 -0.036708076296006364" rpy="0 0 0"/>
    <mass value="0.03309757782688206"/>
    <inertia ixx="7.9e-05" iyy="0.000165" izz="8.7e-05" ixy="0.0" iyz="0.0" ixz="-1e-06"/>
  </inertial>
  <visual>
    <origin xyz="0.042 0.1 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/gear2_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="0.042 0.1 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/gear2_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>

<link name="gear1_1">
  <inertial>
    <origin xyz="-0.04306982066328853 7.806157475943876e-05 -0.03669780485017424" rpy="0 0 0"/>
    <mass value="0.03309757782687909"/>
    <inertia ixx="7.9e-05" iyy="0.000165" izz="8.7e-05" ixy="0.0" iyz="0.0" ixz="-1e-06"/>
  </inertial>
  <visual>
    <origin xyz="-0.041 -0.1 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/gear1_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
    <material name="silver"/>
  </visual>
  <collision>
    <origin xyz="-0.041 -0.1 -0.2" rpy="0 0 0"/>
    <geometry>
      <mesh filename="package://terrabot_design4_description/meshes/gear1_1.stl" scale="0.001 0.001 0.001"/>
    </geometry>
  </collision>
</link>



<!--imu-->  
<link name="imu_link_1">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="red"/>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="imu_joint_1" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0 0 0.215" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="imu_link_1"/>
  </joint>

<link name="imu_link_2">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="red"/>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="imu_joint_2" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0.01 0 0.010" rpy="0 0 0"/>
    <parent link="gear2_1"/>
    <child link="imu_link_2"/>
  </joint>

  <link name="imu_link_3">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="red"/>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

  <joint name="imu_joint_3" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="-0.07 0 0.010" rpy="0 0 0"/>
    <parent link="gear1_1"/>
    <child link="imu_link_3"/>
  </joint>


<joint name="Revolute_55" type="continuous">
  <origin xyz="0.098 0.000205 -0.151" rpy="0 0 0"/>
  <parent link="gear2_1"/>
  <child link="wheel3_1"/>
  <axis xyz="-1.0 0.0 -0.0"/>
</joint>

<joint name="Revolute_56" type="continuous">
  <origin xyz="-0.025 0.000205 -0.151" rpy="0 0 0"/>
  <parent link="gear2_1"/>
  <child link="wheel2_1"/>
  <axis xyz="1.0 -0.0 0.0"/>
</joint>

<joint name="Revolute_57" type="continuous">
  <origin xyz="0.018 0.000321 -0.151" rpy="0 0 0"/>
  <parent link="gear1_1"/>
  <child link="wheel4_1"/>
  <axis xyz="-1.0 0.0 0.0"/>
</joint>

<joint name="Revolute_58" type="continuous">
  <origin xyz="-0.105 0.000321 -0.151" rpy="0 0 0"/>
  <parent link="gear1_1"/>
  <child link="wheel1_1"/>
  <axis xyz="1.0 -0.0 -0.0"/>
</joint>

<joint name="Revolute_59" type="continuous">
  <origin xyz="-0.042 -0.1 0.2" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="gear2_1"/>
  <axis xyz="-1.0 0.0 0.0"/>
</joint>

<joint name="Revolute_60" type="continuous">
  <origin xyz="0.041 0.1 0.2" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="gear1_1"/>
  <axis xyz="-1.0 0.0 0.0"/>
</joint>

</robot>
